src/opendr/perception/semantic_segmentation/bisenet/bisenet_learner.py

```diff
@@ -0,0 +1,309 @@
+        self.crop_height = crop_height
+        self.crop_width = crop_width
+        self.context_path = context_path
+        self.loss_func = torch.nn.CrossEntropyLoss()
+        self.build_model()
+
+        if self.optimizer == "sgd":
+            self.optimizer_func = torch.optim.SGD(self.model.parameters(), lr=self.lr, momentum=momentum)
+        elif self.optimizer == 'rmsprop':
+            self.optimizer_func = torch.optim.RMSprop(self.model.parameters(), lr=self.lr)
+        elif self.optimizer == 'adam':
+            self.optimizer_func = torch.optim.Adam(self.model.parameters(), lr=self.lr)
+
+    def build_model(self):
+        self.model = BiSeNet(num_classes=self.num_classes, context_path=self.context_path)
+        if self.device == 'cuda':
+            self.model = torch.nn.DataParallel(self.model).cuda()
+
+    def fit(self, dataset, val_dataset=None, silent=False, verbose=True):
+        """
+        This method is used for training the algorithm on a train dataset
+
+        """
+
+        dataloader_train = DataLoader(dataset,
+                                      batch_size=self.batch_size,
+                                      shuffle=True,
+                                      num_workers=self.num_workers,
+                                      pin_memory=self.pin_memory,
+                                      drop_last=self.drop_last)
+
+        step = 0
+        for epoch in range(1, self.iters + 1):
+            self.model.train()
+            tq = tqdm.tqdm(total=len(dataloader_train) * self.batch_size)
+            tq.set_description('epoch %d, lr %f' % (epoch, self.lr))
+            loss_record = []
+            for i, (data, label) in enumerate(dataloader_train):
+                if self.device == 'cuda':
+                    data = data.cuda()
+                    label = label.cuda()
+                output, output_sup1, output_sup2 = self.model(data)
+                loss1 = self.loss_func(output, label)
+                loss2 = self.loss_func(output_sup1, label)
+                loss3 = self.loss_func(output_sup2, label)
+                loss = loss1 + loss2 + loss3
+                tq.update(self.batch_size)
+                tq.set_postfix(loss='%.6f' % loss)
+                self.optimizer_func.zero_grad()
+                loss.backward()
+                self.optimizer_func.step()
+                step += 1
+                loss_record.append(loss.item())
+            tq.close()
+            loss_train_mean = np.mean(loss_record)
+            if not silent and verbose:
+                print('Total Loss : %f' % (loss_train_mean))
+            if self.checkpoint_after_iter != 0 and epoch % self.checkpoint_after_iter == 0:
+                self.save(os.path.join(self.temp_path, 'epoch', str(epoch)))
+            if self.val_after != 0 and epoch % self.val_after == 0 and val_dataset is not None:
+                if not silent and verbose:
+                    print('Validation...')
```

Comment:
fit() should return a dictionary with some training statistics according to specs.